Use-time currentcolor resolution in drop-shadow()

anttijk opened the pull request:
#### e55d2e52f72128530e13f647633807cb30e7d4b6
<pre>
Use-time currentcolor resolution in drop-shadow()
<a href="https://bugs.webkit.org/show_bug.cgi?id=290835">https://bugs.webkit.org/show_bug.cgi?id=290835</a>
<a href="https://rdar.apple.com/148317306">rdar://148317306</a>

Reviewed by Sam Weinig.

Fix the last case where we resolve currentcolor at computed-value time.
Besides fixing the inheritance behavior this allows cleaning up various awkward leftovers in the codebase.

To achieve this without a layer violation we add a new FilterOperation type DropShadowFilterOperationWithStyleColor
that lives outside the platform layer. Drop-shadow filters in RenderStyle use this type. It gets resolved and converted to
DropShadowFilterOperation if it needs to passed to the platform layers/IPC.

* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/drop-shadow-currentcolor-inheritance-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/drop-shadow-currentcolor-inheritance.html: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/animation/AnimationUtilities.h:
(WebCore::BlendingContext::BlendingContext):

Include current from/to colors so we can resolve currentcolor everywhere in blending code.

* Source/WebCore/platform/graphics/ca/cocoa/PlatformCAFiltersCocoa.mm:
(WebCore::keyValueCountForFilter):
(WebCore::PlatformCAFilters::presentationModifiers):
(WebCore::PlatformCAFilters::updatePresentationModifiers):
(WebCore::PlatformCAFilters::setFiltersOnLayer):
* Source/WebCore/platform/graphics/filters/FilterOperation.cpp:
(WebCore::DropShadowFilterOperationBase::nonColorEqual const):
(WebCore::DropShadowFilterOperationBase::isIdentity const):
(WebCore::DropShadowFilterOperationBase::outsets const):
(WebCore::DropShadowFilterOperation::operator== const):
(WebCore::DropShadowFilterOperation::blend):
(WebCore::DropShadowFilterOperation::dump const):
(WebCore::operator&lt;&lt;):
(WebCore::DropShadowFilterOperation::isIdentity const): Deleted.
(WebCore::DropShadowFilterOperation::outsets const): Deleted.
* Source/WebCore/platform/graphics/filters/FilterOperation.h:
(WebCore::FilterOperation::isDropShadowBase const):
(WebCore::FilterOperation::requiresRepaintForCurrentColorChange const):
* Source/WebCore/platform/graphics/filters/FilterOperations.cpp:
(WebCore::FilterOperations::requiresRepaintForCurrentColorChange const):
* Source/WebCore/platform/graphics/filters/FilterOperations.h:
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::createDropShadowEffect):
(WebCore::CSSFilter::buildFilterFunctions):

Map to DropShadowFilterOperationWithStyleColor

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateFiltersAfterStyleChange):

Reset cached filter on currentcolor change if needed.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::resolveFilters):
(WebCore::RenderLayerBacking::updateFilters):
(WebCore::RenderLayerBacking::updateBackdropFilters):

Convert to DropShadowFilterOperation if needed.

* Source/WebCore/rendering/style/DropShadowFilterOperationWithStyleColor.cpp: Added.
(WebCore::Style::DropShadowFilterOperationWithStyleColor::createEquivalentWithResolvedColor):
(WebCore::Style::DropShadowFilterOperationWithStyleColor::operator== const):
(WebCore::Style::DropShadowFilterOperationWithStyleColor::blend):
(WebCore::Style::DropShadowFilterOperationWithStyleColor::dump const):
* Source/WebCore/rendering/style/DropShadowFilterOperationWithStyleColor.h: Added.
(WebCore::Style::DropShadowFilterOperationWithStyleColor::create):
(WebCore::Style::DropShadowFilterOperationWithStyleColor::styleColor const):
(WebCore::Style::DropShadowFilterOperationWithStyleColor::DropShadowFilterOperationWithStyleColor):

The new style-only filter operation type is in the Style namespace.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::changeRequiresLayerRepaint const):
* Source/WebCore/style/StyleInterpolation.cpp:
(WebCore::Style::Interpolation::interpolateStandardProperty):
(WebCore::Style::Interpolation::interpolateCustomProperty):
* Source/WebCore/style/StyleInterpolationContext.h:
(WebCore::Style::Interpolation::Context::Context):
* Source/WebCore/style/StyleInterpolationFunctions.h:
(WebCore::Style::Interpolation::blendFunc):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::applyMatchedProperties):

Hack not needed anymore.

* Source/WebCore/style/values/color/StyleColor.cpp:
(WebCore::Style::toStyleColorWithResolvedCurrentColor): Deleted.

Hack not needed anymore.

* Source/WebCore/style/values/color/StyleColor.h:
* Source/WebCore/style/values/filter-effects/StyleDropShadowFunction.cpp:
(WebCore::Style::toCSSDropShadow):
(WebCore::Style::createFilterOperation):
* Source/WebCore/style/values/filter-effects/StyleDropShadowFunction.h:
* Source/WebCore/style/values/filter-effects/StyleFilterProperty.cpp:
(WebCore::Style::toCSSFilterProperty):

Canonical link: <a href="https://commits.webkit.org/293027@main">https://commits.webkit.org/293027@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fcba6ded386b26536e15aa5ba8aa453e7f143ecd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97721 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17346 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7563 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102862 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48250 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99766 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17640 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25799 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/74456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31620 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100724 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13371 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88337 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/54808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13141 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6253 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47693 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83142 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6330 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104829 "Hash fcba6ded for PR 43380 does not build (failure)") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24801 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18074 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/104829 "Hash fcba6ded for PR 43380 does not build (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25173 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84497 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/104829 "Hash fcba6ded for PR 43380 does not build (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20887 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27483 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5163 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18408 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24762 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/29931 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24584 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27898 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26158 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
Do we progress LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/drop-shadow-currentcolor-dynamic-002.html ?

weinig reviewed (approved):
Nice!

I think this is a good improvement, but we probably need to have a Style level version of the FilterOperations at some point to fix all the layering violation in this code. The accelerated animations make that tricky though.

anttijk committed:
Use-time currentcolor resolution in drop-shadow()
https://bugs.webkit.org/show_bug.cgi?id=290835
rdar://148317306

Reviewed by Sam Weinig.

Fix the last case where we resolve currentcolor at computed-value time.
Besides fixing the inheritance behavior this allows cleaning up various awkward leftovers in the codebase.

To achieve this without a layer violation we add a new FilterOperation type DropShadowFilterOperationWithStyleColor
that lives outside the platform layer. Drop-shadow filters in RenderStyle use this type. It gets resolved and converted to
DropShadowFilterOperation if it needs to passed to the platform layers/IPC.

* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/drop-shadow-currentcolor-inheritance-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/filter-effects/drop-shadow-currentcolor-inheritance.html: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/animation/AnimationUtilities.h:
(WebCore::BlendingContext::BlendingContext):

Include current from/to colors so we can resolve currentcolor everywhere in blending code.

* Source/WebCore/platform/graphics/ca/cocoa/PlatformCAFiltersCocoa.mm:
(WebCore::keyValueCountForFilter):
(WebCore::PlatformCAFilters::presentationModifiers):
(WebCore::PlatformCAFilters::updatePresentationModifiers):
(WebCore::PlatformCAFilters::setFiltersOnLayer):
* Source/WebCore/platform/graphics/filters/FilterOperation.cpp:
(WebCore::DropShadowFilterOperationBase::nonColorEqual const):
(WebCore::DropShadowFilterOperationBase::isIdentity const):
(WebCore::DropShadowFilterOperationBase::outsets const):
(WebCore::DropShadowFilterOperation::operator== const):
(WebCore::DropShadowFilterOperation::blend):
(WebCore::DropShadowFilterOperation::dump const):
(WebCore::operator<<):
(WebCore::DropShadowFilterOperation::isIdentity const): Deleted.
(WebCore::DropShadowFilterOperation::outsets const): Deleted.
* Source/WebCore/platform/graphics/filters/FilterOperation.h:
(WebCore::FilterOperation::isDropShadowBase const):
(WebCore::FilterOperation::requiresRepaintForCurrentColorChange const):
* Source/WebCore/platform/graphics/filters/FilterOperations.cpp:
(WebCore::FilterOperations::requiresRepaintForCurrentColorChange const):
* Source/WebCore/platform/graphics/filters/FilterOperations.h:
* Source/WebCore/rendering/CSSFilter.cpp:
(WebCore::createDropShadowEffect):
(WebCore::CSSFilter::buildFilterFunctions):

Map to DropShadowFilterOperationWithStyleColor

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateFiltersAfterStyleChange):

Reset cached filter on currentcolor change if needed.

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::resolveFilters):
(WebCore::RenderLayerBacking::updateFilters):
(WebCore::RenderLayerBacking::updateBackdropFilters):

Convert to DropShadowFilterOperation if needed.

* Source/WebCore/rendering/style/DropShadowFilterOperationWithStyleColor.cpp: Added.
(WebCore::Style::DropShadowFilterOperationWithStyleColor::createEquivalentWithResolvedColor):
(WebCore::Style::DropShadowFilterOperationWithStyleColor::operator== const):
(WebCore::Style::DropShadowFilterOperationWithStyleColor::blend):
(WebCore::Style::DropShadowFilterOperationWithStyleColor::dump const):
* Source/WebCore/rendering/style/DropShadowFilterOperationWithStyleColor.h: Added.
(WebCore::Style::DropShadowFilterOperationWithStyleColor::create):
(WebCore::Style::DropShadowFilterOperationWithStyleColor::styleColor const):
(WebCore::Style::DropShadowFilterOperationWithStyleColor::DropShadowFilterOperationWithStyleColor):

The new style-only filter operation type is in the Style namespace.

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::changeRequiresLayerRepaint const):
* Source/WebCore/style/StyleInterpolation.cpp:
(WebCore::Style::Interpolation::interpolateStandardProperty):
(WebCore::Style::Interpolation::interpolateCustomProperty):
* Source/WebCore/style/StyleInterpolationContext.h:
(WebCore::Style::Interpolation::Context::Context):
* Source/WebCore/style/StyleInterpolationFunctions.h:
(WebCore::Style::Interpolation::blendFunc):
* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::applyMatchedProperties):

Hack not needed anymore.

* Source/WebCore/style/values/color/StyleColor.cpp:
(WebCore::Style::toStyleColorWithResolvedCurrentColor): Deleted.

Hack not needed anymore.

* Source/WebCore/style/values/color/StyleColor.h:
* Source/WebCore/style/values/filter-effects/StyleDropShadowFunction.cpp:
(WebCore::Style::toCSSDropShadow):
(WebCore::Style::createFilterOperation):
* Source/WebCore/style/values/filter-effects/StyleDropShadowFunction.h:
* Source/WebCore/style/values/filter-effects/StyleFilterProperty.cpp:
(WebCore::Style::toCSSFilterProperty):

Canonical link: https://commits.webkit.org/293027@main
